src/services/utilities.ts

```diff
@@ -1448,6 +1464,12 @@ namespace ts {
 
     // Get info for an expression like `f <` that may be the start of type arguments.
     export function getPossibleTypeArgumentsInfo(tokenIn: Node, sourceFile: SourceFile): PossibleTypeArgumentInfo | undefined {
+        // This is a rare case, but one that saves on a _lot_ of work if true - if the sorce file has _no_ `<` character,
+        // then there obviously can't be any type arguments - no expensive bracematching backwards scanning required
+        if (sourceFile.text.indexOf("<") === -1) {

```

Comment:
However this fairly innocuous and obvious-in-hindsight line takes the cake, and is what essentially eliminated most of the test's runtime. As the comment says: if there's no `<` in the file, why should we brace match anything at all, when we can trivially see we can't be in a type position? So by realizing that here, we skip the whole process, making the completions request return pretty much instantly.

Reply:
How much improvement do you see on `excessivelyLargeArrayLiteralCompletions.ts` if it has a `<` somewhere in it?

Reply:
Sorry, you answered that in the other location—36s to 10s 👍

Reply:
Can you limit this from looking at the entire `sourceFile`?

Reply:
Only a `sourceFile` has text?

Reply:
but you can use lastIndexOf with start position as current position ?

Reply:
Also, you have a node which sounds like it might contain the range of text you're interested in

Reply:
Ohh, you mean only searching from the start of the file to the start of the input token, yeah, that's fair.